Firestore/core/test/firebase/firestore/util/string_win_test.cc

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/string_win.h"
+
+#include "absl/strings/string_view.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+#if defined(_WIN32)
+
+TEST(StringWindowsTest, Empty) {
+  EXPECT_EQ(std::wstring{L""}, Utf8ToNative(""));
+  EXPECT_EQ(std::string{""}, NativeToUtf8(L""));
+}
+
+TEST(StringWindowsTest, EmbeddedNulls) {
+  std::string embedded_nulls({'\0', ' ', 'a'}, 3);
+  std::wstring wembedded_nulls({L'\0', L' ', L'a'}, 3);
+  EXPECT_EQ(wembedded_nulls, Utf8ToNative(embedded_nulls));
+  EXPECT_EQ(embedded_nulls, NativeToUtf8(wembedded_nulls));
+}
+
+TEST(StringWindowsTest, NonAscii) {
+  // left and right curly quotation marks
+  std::string curly{u8"\u2018hi\u2019"};
+  std::wstring wcurly{L"\u2018hi\u2019"};
+  EXPECT_EQ(wcurly, Utf8ToNative(curly));
+  EXPECT_EQ(curly, NativeToUtf8(wcurly));
+}
+
+TEST(StringWindowsTest, InvalidUtf8) {
+  // The 0xFF byte is not valid in UTF-8; Windows will replace with U+FFFD
+  // (replacement character)
+  std::string invalid{"\xff\xff"};
+  std::wstring winvalid_replaced{L"\ufffd\ufffd"};
+  EXPECT_EQ(winvalid_replaced, Utf8ToNative(invalid));
+
+  // Missing the trailing part of the surrogate pair. MSVC seemingly won't allow
+  // surrogates to appear in a string literal (complete or otherwise).
+  std::wstring winvalid(L"AA");
+  winvalid[1] = static_cast<wchar_t>(0xD800);
+  std::string invalid_replacement{u8"A\ufffd"};
+  EXPECT_EQ(invalid_replacement, NativeToUtf8(winvalid));
+}
+
+TEST(StringWindowsTest, LastErrorMessage) {
+  EXPECT_EQ(std::string{"The parameter is incorrect."}, LastErrorMessage(0x57));
```

Comment:
FormatMessageW's docs are pretty vague about what the language id does. But https://msdn.microsoft.com/en-us/library/ms534732(VS.85).aspx implies "LANG_NEUTRAL, which is the user's default language."

I suspect this means that if you set your language to something other than english, this test will break.

Reply:
I've implemented a check that prevents a spurious test failure if your primary language identifier is non-English, though I'm not sure what other test to apply in that case. Ideas?

Reply:
Some possible suggestions. Feel free to do any/all/none. (But if none, do note that message templates can change even within a single language. See (2) for details. IOW, your test could still spuriously fail even on en systems, though I suspect that would be unlikely.)

1) Non-empty string? Getting more fancy, you could ensure it *doesnt* look like "error %s; unknown error %s while getting error text" and I suppose also doesn't end in \r\n.

2) Another alternative: Force the language to en-us. My understanding is that that should be present on all Windows systems. (The test would still fail if MS decided to change the string message template. I believe they *do* do that from time to time. And perhaps worse; they also can vary by OS or patch level, so the test, even as currently written, might fail on some systems.) That's unnecessarily impolite i18n, but OTOH, to date, all of our developer facing error messages are in english. Maybe there's a consistency argument to be made? Especially since the HARD_ASSERT statement prefixes the error message with english text.

If you go this route, and want to protect yourself from the whims of MS's tech writers, you could search (case insensitive) the resulting error string for keywords that are unlikely to be removed. "parameter" might be a good one in this case. It's not a guarantee, but it should help. Well known/old error messages like this one seem less likely to change, which should also help.

3) Yet another alternative: You could override LastErrMsg to take in a language id (defaulting to LANG_NEUTRAL). Then the tests could force en-us (since it's supposed to always be present.) Perhaps that's overkill.

Testing error messages is often brittle. Best I've seen is some variant of (1), then ensure that the method in question is as small/trivial as possible, and finally do all other error checking via error codes or caught exceptions. (i.e. compare numbers/types rather than language strings.)

Reply:
I went with (2), which seems easiest. If they ever change this specific text I'll:
  * be surprised and
  * suck it up and figure out how to implement (1).